From 9793e5715821b1aea89e1cf8e32863947dddfb0d Mon Sep 17 00:00:00 2001
From: Wenzong Fan <wenzong.fan@windriver.com>
Date: Thu, 4 Jan 2018 05:11:36 +0000
Subject: [PATCH] Fix error for spawning /lib/systemd/systemd

Port patches from https://github.com/fedora-selinux/selinux-policy:
  cf08e3c1 Add init_entrypoint_exec() interface.
  49a30cfe Allow user domains with login_userdomain to have entrypoint \
           access on init_exec. It is needed by pam_selinux.so call in \
           systemd-users. BZ(#1263350)
  Pick fixes from:
  32ec69ac Remove init_systemd and init_upstart boolean, Move \
           init_daemon_domain and init_system_domain to use attributes \
           to shrink policy
  3eaa9939 UPdate for f14 policy

Remove error:
  user@0.service: Failed at step EXEC spawning \
  /lib/systemd/systemd: Permission denied

Fix avc denials:
  avc: denied { sendto } for pid=567 comm="systemd" \
  path="/run/systemd/notify" \
  scontext=root:sysadm_r:sysadm_t:s0 \
  tcontext=system_u:system_r:init_t:s0-s15:c0.c1023 \
  tclass=unix_dgram_socket

  avc: denied { create } for pid=567 comm="systemd" \
  scontext=root:sysadm_r:sysadm_t:s0 \
  tcontext=root:sysadm_r:sysadm_t:s0 \
  tclass=netlink_kobject_uevent_socket

Upstream-Status: Pending

Signed-off-by: Wenzong Fan <wenzong.fan@windriver.com>
---
 policy/modules/system/init.if       | 18 ++++++++++++++++++
 policy/modules/system/userdomain.if |  5 +++++
 2 files changed, 23 insertions(+)

diff --git a/policy/modules/system/init.if b/policy/modules/system/init.if
index 2ebf9f7..9face27 100644
--- a/policy/modules/system/init.if
+++ b/policy/modules/system/init.if
@@ -651,6 +651,24 @@ interface(`init_domtrans',`
 
 ########################################
 ## <summary>
+##	Allow any file point to be the entrypoint of this domain.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`init_entrypoint_exec',`
+	gen_require(`
+		type init_exec_t;
+	')
+
+	allow $1 init_exec_t:file entrypoint;
+')
+
+########################################
+## <summary>
 ##	Execute the init program in the caller domain.
 ## </summary>
 ## <param name="domain">
diff --git a/policy/modules/system/userdomain.if b/policy/modules/system/userdomain.if
index 5a28862..4c4ffa5 100644
--- a/policy/modules/system/userdomain.if
+++ b/policy/modules/system/userdomain.if
@@ -610,6 +610,8 @@ template(`userdom_common_user_template',`
 	# gnome-settings-daemon and some applications create a netlink socket
 	allow $1_t self:netlink_kobject_uevent_socket create_socket_perms;
 
+	allow $1_t self:socket create_socket_perms;
+
 	allow $1_t unpriv_userdomain:fd use;
 
 	kernel_read_system_state($1_t)
@@ -981,6 +983,9 @@ template(`userdom_login_user_template', `
 	init_dontaudit_use_fds($1_t)
 	init_dontaudit_use_script_fds($1_t)
 
+	# Needed by pam_selinux.so calling in systemd-users
+	init_entrypoint_exec(login_userdomain)
+
 	libs_exec_lib_files($1_t)
 
 	logging_dontaudit_getattr_all_logs($1_t)
-- 
2.7.4

